### PR TITLE
[Feature] Adds ordinal distributions

### PR DESCRIPTION
**Pull request by louisfaury**
## Description

Adds an ordinally parametrised distribution from [[Tang & Agrawal, 2020]](https://arxiv.org/pdf/1901.10500).

## Motivation and Context

This parametrisation showed useful when learning distribution on finite sets that were obtained by discretising continuous sets. 

Note: We can provide an example, although it should probably live in the `torchrl` repository. 

## Types of changes

What types of changes does your code introduce? Remove all that do not apply:

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds core functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] Documentation (update in the documentation)
- [ ] Example (update in the folder of examples)

## Checklist

Go over all the following points, and put an `x` in all the boxes that apply.
If you are unsure about any of these, don't hesitate to ask. We are here to help!

- [x] I have read the [CONTRIBUTION](https://github.com/pytorch/tensordict/blob/main/CONTRIBUTING.md) guide (**required**)
- [x] My change requires a change to the documentation.
- [x] I have updated the tests accordingly (*required for a bug fix or a new feature*).
- [x] I have updated the documentation accordingly.

### CODE REVIEW

File: torchrl/modules/distributions/discrete.py [hunk 1]
```diff
@@ -56,7 +54,7 @@ class ReparamGradientStrategy(Enum):
 class OneHotCategorical(D.Categorical):
     """One-hot categorical distribution.
 
-    This class behaves excacly as torch.distributions.Categorical except that it reads and produces one-hot encodings
+    This class behaves exactly as torch.distributions.Categorical except that it reads and produces one-hot encodings
```
Comment (louisfaury) [inline torchrl/modules/distributions/discrete.py:57]: Took the opportunity to address a few typos in this file.

File: torchrl/modules/distributions/discrete.py [hunk 2]
```diff
@@ -469,3 +467,57 @@ def rsample(self, sample_shape: Union[torch.Size, Sequence] = None) -> torch.Ten
...
+        scores (torch.Tensor): a tensor of shape [..., N] where N is the size of the set which supports the distributions.
+            Typically, the output of a neural network parametrising the distribution.
+    """
+
+    def __init__(self, scores: torch.Tensor):
+        logits = _generate_ordinal_logits(scores)
+        super().__init__(logits=logits)
+
+
+def _generate_ordinal_logits(scores: torch.Tensor) -> torch.Tensor:
```
Comment (louisfaury) [inline torchrl/modules/distributions/discrete.py:508]: Perhaps we could make this a static method of `Ordinal`, depending on preferences.

File: torchrl/modules/distributions/discrete.py [hunk 3]
```diff
@@ -66,7 +64,7 @@ class OneHotCategorical(D.Categorical):
             reparameterized samples.
             ``ReparamGradientStrategy.PassThrough`` will compute the sample gradients
              by using the softmax valued log-probability as a proxy to the
-             samples gradients.
+             sample gradients.
```
Comment (louisfaury) [inline torchrl/modules/distributions/discrete.py:67]: Note: I _think_ the `num_params` class attribute defined at l. 82/84 is not used. If that's correct, I can take the opportunity to remove this in this MR.

File: torchrl/modules/distributions/discrete.py [hunk 4]
```diff
@@ -469,3 +467,57 @@ def rsample(self, sample_shape: Union[torch.Size, Sequence] = None) -> torch.Ten
             raise ValueError(
                 f"Unknown reparametrization strategy {self.reparam_strategy}."
             )
+
+
+class Ordinal(D.Categorical):
```
Comment (louisfaury) [inline torchrl/modules/distributions/discrete.py:472]: Note: we _could_ have a masked version of this distribution (although it is rare to have masks/illegal actions for continuously supported distributions -- hence since this is used after a discretisation layer, the usefulness of supporting masks is imo arguable). Anyway, would leave it FAR.

File: torchrl/modules/distributions/discrete.py [hunk 5]
```diff
@@ -469,3 +467,57 @@ def rsample(self, sample_shape: Union[torch.Size, Sequence] = None) -> torch.Ten
...
+    useful for learning discrete sampling from continuous sets. See §5 of
+    `Tang & Agrawal, 2020<https://arxiv.org/pdf/1901.10500.pdf>`_ for details.
+
+    .. note::
+        This class is mostly useful when you want to learn a distribution over
+        a finite set which is obtained by discretising a continuous set.
+
+    Args:
+        scores (torch.Tensor): a tensor of shape [..., N] where N is the size of the set which supports the distributions.
+            Typically, the output of a neural network parametrising the distribution.
```
Comment (vmoens) [inline torchrl/modules/distributions/discrete.py:487]: If you could add a simple example it'd be awesome, what about (feel free to edit)

```suggestion
            Typically, the output of a neural network parametrising the distribution.

    Examples:
        >>> batch_size = 1
        >>> seq_size = 4
        >>> num_categories = 3
        >>> logits = torch.ones((batch_size, seq_size, num_categories))
        >>> sampler = Ordinal(scores=logits)
        >>> # The probabilities and logits are not equal anymore, but reflect the ordering of the samples
        >>> sampler.probs
        tensor([[[0.0900, 0.2447, 0.6652],
                 [0.0900, 0.2447, 0.6652],
                 [0.0900, 0.2447, 0.6652],
                 [0.0900, 0.2447, 0.6652]]])
        >>> torch.manual_seed(10)
        >>> # We print an histogram to reflect the sampling frequency of each item (0, 1 and 2)
        >>> torch.histogram(sampler.sample((1000,)).reshape(-1).float(), bins=3)
        torch.return_types.histogram(
            hist=tensor([ 386.,  989., 2625.]),
            bin_edges=tensor([0.0000, 0.6667, 1.3333, 2.0000]))
```

**Review by vmoens — APPROVED** [review]
LGTM thanks for this! 
See the couple of suggestions but otherwise good to go

File: torchrl/modules/distributions/discrete.py (hunk 5 above)
Comment (louisfaury) [reply torchrl/modules/distributions/discrete.py:487]: I thought about this but I was unsure about its meaningfulness -- since this is a "learning-time" feature, here we wouldn't be showcasing anything that cannot be done by a Categorical distribution. 

I thought about it and wrote an example which involves some learning, maybe it is more useful like that. Wdyt ?

Comment (louisfaury) [reply torchrl/modules/distributions/discrete.py:485]: Agreed with all of the above; do you think the one I pushed this morning (copied below for visibility) clashes with 2. and 3 ? 

```python
"""
    Examples:
        >>> num_atoms, num_samples = 5, 20
        >>> mean = (num_atoms - 1) / 2  # Target mean for samples, centered around the middle atom
        >>> torch.manual_seed(42)
        >>> logits = torch.ones((num_atoms), requires_grad=True)
        >>> optimizer = torch.optim.Adam([logits], lr=0.1)
        >>>
        >>> # Perform optimisation loop to minimise deviation from `mean`
        >>> for _ in range(20):
        >>>     sampler = Ordinal(scores=logits)
        >>>     samples = sampler.sample((num_samples,))
        >>>     # Define loss to encourage samples around the mean by penalising deviation from mean
        >>>     loss = torch.mean((samples - mean) ** 2 * sampler.log_prob(samples))
        >>>     loss.backward()
        >>>     optimizer.step()
        >>>     optimizer.zero_grad()
        >>>
        >>> sampler.probs
        tensor([0.0308, 0.1586, 0.4727, 0.2260, 0.1120], ...)
        >>> # Print histogram to observe sample distribution frequency across 5 bins (0, 1, 2, 3, and 4)
        >>> torch.histogram(sampler.sample((1000,)).reshape(-1).float(), bins=num_atoms)
        torch.return_types.histogram(
            hist=tensor([ 24., 158., 478., 228., 112.]),
            bin_edges=tensor([0.0000, 0.8000, 1.6000, 2.4000, 3.2000, 4.0000]))
    """
```

**Review by vmoens — APPROVED** [review]
